Update integration notebook

sjspielman opened the pull request:
Closes #644 
Closes #651 
Closes #657 

This PR updates the integration notebook:
- #644 fixes change the fastMNN bioc link to from `3.16` ->  `release`
- #651 are the roadmap diagrams. I figure we can use this PR to tweak some of the phrasing in those we've been unsure about. Any thoughts on whether these are appearing the the right place in the notebook? I went back and forth for putting the integration roadmap at the top vs. in the integration section..for now at the top!
- #657 modifies how we introduce `purrr`.  One aspect of this I realized I wasn't sure about was whether to remove "vector" as a datatype for `map` input.. At first I was going to remove it to go hard on lists, but we actually do pass a vector in for the first real call to `map`! I tried something out here as a start to finding the right middle ground, feedback welcome!

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -89,45 +93,39 @@ We could also use a `for` loop, which is the approach that many programming lang
-
 The output from running `purr::map()` is always a list (but note that there are other `purrr::map()` relatives which return other object types, as you can read about in [the `purrr::map()` documentation](https://purrr.tidyverse.org/reference/index.html)).
+If this concept sounds a little familiar to you, that's because it probably is!
+In fact, this is conceptually the same as base R's `lapply` function, and the `purrr::map()` family of functions provide more robust and reliable alternatives to the base R `apply` functions.
```
I'm trying to hedge this a bit... I don't think `purrr` provides a good alternative to `apply()` itself!

```suggestion
Base R's `lapply()` function can provide similar utility, and the `purrr::map()` family of functions can in part be thought of as an alternative to some of the base R `apply` functions, with more consistent behavior.
```

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -177,7 +175,10 @@ sce_paths
-Since `readr::read_rds()` can only operate on one input at a time (unlike `sqrt()` which can operate across a vector), we'll need to use `purrr::map()` to run it on all input file paths in one command.
+Since `readr::read_rds()` can only operate on one input at a time, we'll need to use `purrr::map()` to run it on all input file paths in one command.
+Although `sce_paths` is a vector (not a list), it will still work as input to `purrr:map()`.
+But, the output from this code will be a list, since that's what `purrr::map()` always returns.
+(We could convert `sce_paths` to a list first with `as.list(sce_paths)`, but for this example, the code will work just the same!)
```
I don't think adding a suggestion of something not to do is particularly helpful here. I also don't like starting with a "but" if we don't really need to!
```suggestion
The output from this code will still be a list, since that's what `purrr::map()` always returns.
```

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -217,6 +218,9 @@ If we have annotations, they are a helpful "bonus" for assessing the integration
 
 ## Prepare the SCE list for integration
 
+![Single-cell roadmap: Merge](diagrams/roadmap_multi_merge.png)
```
Whether we need the merge diagram here again is up to you. I don't think it is terrible to have it twice, but we probably don't need to, as we will have a second integration diagram later. Or if you take the suggestion to change the initial diagram, leaving this here makes sense!

jashapiro reviewed (commented):
This looks like a good start. I like what you did with the purrr content; I wonder if we might also want to make the `glue` content non-live the first time through? More importantly though, I realized that we really should separate the `sample_names` vector from the file path creation: we really don't want people to be typing in those sample names live.

> Any thoughts on whether these are appearing the the right place in the notebook? I went back and forth for putting the integration roadmap at the top vs. in the integration section..for now at the top!

I left some comments inline but the main thought is that roadmap should appear in the order of the highlighted boxes, which means that merge should definitely come before integration, especially since you don't want to be explaining the integration substeps before you get there in the notebook. I suggested an alternative of one more diagram with all the steps for this notebook highlighted, and I think that is what I like the best (despite the code suggestion).  What do you think of that idea?

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -27,14 +27,18 @@ Specifically, we'll have a look at four samples from the [`SCPCP000005` project]
 Regardless of which integration tool is used, the `SingleCellExperiment` (SCE) objects first need to be reformatted and merged into a single (uncorrected!) SCE object that contains all cells from all samples.
 This merged SCE can then be used for integration to obtain a formally batch-corrected SCE object.
 
+![Single-cell roadmap: Integrate](diagrams/roadmap_multi_integrate.png)
```
I had a second, so here is a version of the diagram I am suggesting: https://docs.google.com/presentation/d/1j_qH4KD2WxAbwX3n9vCkoiG87E1wIhnYzThjk8VgZ34/edit#slide=id.g21270a3c171_0_197

sjspielman commented:
Changes since last review:
- New over diagram at the top! https://github.com/AlexsLemonade/training-modules/pull/662/commits/b83a62055da9f86165d1e57f02c93b55764efe03
- Unlived the `sample_names` (and collapsed a print chunk into it, we don't need that many chunks). Inspired by this change, I also unlivened the `case_when()` chunk which is also a typo trap. https://github.com/AlexsLemonade/training-modules/pull/662/commits/a9c7c0caeabdcd3ed88bba9a0018443ee98f71c5
- I agree that it makes sense to unliven `glue` as well, at least for virtual and see how it goes! https://github.com/AlexsLemonade/training-modules/pull/662/commits/38cdd2f9099414c209d3306e03619615aafbd79d

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -89,45 +93,39 @@ We could also use a `for` loop, which is the approach that many programming lang
-           <function to apply to each item in the input>, 
-           <any additional arguments to the function can go here>, 
-           <and also here if there are even more arguments, and so on>)
+purrr::map(input list or vector, 
+           function to apply to each item in the input, 
+           any additional arguments to the function can go here, 
+           and also here if there are even more arguments, and so on)
```
I didn't notice this change, but I would probably revert the removal of the angle brackets. Having the bare text seems confusing to me.

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -89,45 +93,39 @@ We could also use a `for` loop, which is the approach that many programming lang
-
 The output from running `purr::map()` is always a list (but note that there are other `purrr::map()` relatives which return other object types, as you can read about in [the `purrr::map()` documentation](https://purrr.tidyverse.org/reference/index.html)).
+If this concept sounds a little familiar to you, that's because it probably is!
+Base R's `lapply()` function can provide similar utility, and the `purrr::map()` family of functions can in part be thought of as an alternative to some of the base R `apply` functions, with more consistent behavior.
```
Making "in part" parenthetical
```suggestion
Base R's `lapply()` function can provide similar utility, and the `purrr::map()` family of functions can (in part) be thought of as an alternative to some of the base R `apply` functions, with more consistent behavior.
```

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -27,14 +27,18 @@ Specifically, we'll have a look at four samples from the [`SCPCP000005` project]
 Regardless of which integration tool is used, the `SingleCellExperiment` (SCE) objects first need to be reformatted and merged into a single (uncorrected!) SCE object that contains all cells from all samples.
 This merged SCE can then be used for integration to obtain a formally batch-corrected SCE object.
 
+![Single-cell roadmap: Integration Overview](diagrams/roadmap_multi_merge-integrate.png)
```
With this change, the other part of my comment seems to have been hidden on this page, so I am adding it again here:

> In conjunction with this diagram (either version), I think it will be worth adding a note about what processing has already been done on these samples. i.e., that they have already been filtered and normalized.

I also think we could now move this diagram up higher in this notebook...after the first paragraph probably makes sense, and that would be a good place to also add a bit more text about ScPCA processing that has been done.

jashapiro reviewed (commented):
Okay, looks pretty good with the three figures now. I just a few more little changes to add & one thing to partially undo...

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -22,19 +22,23 @@ This notebook will demonstrate how to:
 
 In this notebook, we'll perform integration on scRNA-seq datasets from the [Single-cell Pediatric Cancer Atlas (`ScPCA`)](https://scpca.alexslemonade.org/), a database of uniformly-processed pediatric scRNA-seq data built and maintained by the Data Lab.
 This database currently hosts projects from ten different ALSF-funded labs who have generated single-cell pediatric cancer transcriptomic data with the ultimate goal of making this data easily accessible to investigators (like you!).
+**MORE TEXT HERE ON SCPCA PROCESSING**
```
Just to be clear, this can be a sentence that says the data was mapped with alevin fry, then filtered and normalized... If you want to link to the docs, you can do that, but I don't think we need much detail at all!

sjspielman commented:
ScPCA details (too little? too much? goldilocks?) have been added in https://github.com/AlexsLemonade/training-modules/pull/662/commits/a141f78e9c040e4802d1655409a9b7b6ac2b613d!

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -21,20 +21,26 @@ This notebook will demonstrate how to:
 
 In this notebook, we'll perform integration on scRNA-seq datasets from the [Single-cell Pediatric Cancer Atlas (`ScPCA`)](https://scpca.alexslemonade.org/), a database of uniformly-processed pediatric scRNA-seq data built and maintained by the Data Lab.
-This database currently hosts projects from ten different ALSF-funded labs who have generated single-cell pediatric cancer transcriptomic data with the ultimate goal of making this data easily accessible to investigators (like you!).
+This database currently hosts (as of March 2023) eleven projects from eight different ALSF-funded labs who have generated single-cell pediatric cancer transcriptomic data with the ultimate goal of making this data easily accessible to investigators (like you!).
+`ScPCA` raw data is processed with [nextflow pipeline](https://github.com/AlexsLemonade/scpca-nf) that uses [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3) for fast and efficient mapping, followed by filtering and normalization with Bioconductor tools. 
+You can read more about how data in the `ScPCA` is processed in [the associated documentation](https://scpca.readthedocs.io/en/latest/).
```
I was really looking for both more and less here... I was mostly hoping to just get across that we used the most of same tools that they have seen in the first notebook. I added some of that detail below, but check that it is correct!

I also don't think the stats are needed here, especially if we have to keep updating them, so I went the opposite direction and cut the numbers. I also realized the "ultimate goal" part was ambiguous (who's goal?). So I tried to rephrase a bit.

```suggestion
The `ScPCA` database currently hosts single-cell pediatric cancer transcriptomic data generated by ALSF-funded labs, with the ultimate goal of making this data easily accessible to investigators (like you!).
The expression data in `ScPCA` were mapping and quantified with [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3), followed by processing with Bioconductor tools, following the same general procedures that we have covered in this workshop.
The raw counts matrix was filtered with `emptyDropsCellRanger()` and `miQC`, log-normalized with `scuttle`, with dimension reduction by `scater`.
The processed data is stored as `.rds` files containing `SingleCellExperiment` objects.
You can read more about how data in the `ScPCA` is processed in [the associated documentation](https://scpca.readthedocs.io/en/latest/).
```

sjspielman commented:
@jashapiro I did a couple phrasing updates, including "ultimate goal" --> "a leading goal" - https://github.com/AlexsLemonade/training-modules/pull/662/commits/21e54e22ffb8dcca0cd21c20987e59d75cca2ff9

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -21,14 +21,16 @@ This notebook will demonstrate how to:
-`ScPCA` raw data is processed with [nextflow pipeline](https://github.com/AlexsLemonade/scpca-nf) that uses [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3) for fast and efficient mapping, followed by filtering and normalization with Bioconductor tools. 
+The `ScPCA` database currently hosts single-cell pediatric cancer transcriptomic data generated by ALSF-funded labs, with a leading goal of making this data easily accessible to investigators (like you!).
+The expression data in `ScPCA` were mapping and quantified with [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3), followed by processing with Bioconductor tools using the same general procedures that we have covered in this workshop.
+The raw counts matrix was filtered with `emptyDropsCellRanger()` and `miQC` and log-normalized with `scuttle`, with dimension reduction performed by `scater`.
```
and -> Comma
```suggestion
The raw counts matrix was filtered with `emptyDropsCellRanger()` and `miQC`, log-normalized with `scuttle`, with dimension reduction performed by `scater`.
```

jashapiro commented on scRNA-seq-advanced/03-dataset_integration.Rmd:
```diff
@@ -21,14 +21,16 @@ This notebook will demonstrate how to:
 In this notebook, we'll perform integration on scRNA-seq datasets from the [Single-cell Pediatric Cancer Atlas (`ScPCA`)](https://scpca.alexslemonade.org/), a database of uniformly-processed pediatric scRNA-seq data built and maintained by the Data Lab.
-This database currently hosts (as of March 2023) eleven projects from eight different ALSF-funded labs who have generated single-cell pediatric cancer transcriptomic data with the ultimate goal of making this data easily accessible to investigators (like you!).
-`ScPCA` raw data is processed with [nextflow pipeline](https://github.com/AlexsLemonade/scpca-nf) that uses [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3) for fast and efficient mapping, followed by filtering and normalization with Bioconductor tools. 
+The `ScPCA` database currently hosts single-cell pediatric cancer transcriptomic data generated by ALSF-funded labs, with a leading goal of making this data easily accessible to investigators (like you!).
```
I don't love "leading goal"... "primary goal", "main goal"? None of those sound quite right. I have "the" goal below, because just "a goal" sounded bad. 
```suggestion
The `ScPCA` database currently hosts single-cell pediatric cancer transcriptomic data generated by ALSF-funded labs, with the goal of making this data easily accessible to investigators (like you!).
```

jashapiro reviewed (approved):
LGTM! I made a couple more tiny comments/suggestions, but I don't need to approve your ultimate decision on them!

sjspielman commented:
> and -> Comma

Only if it's a compound sentence! Since the subject is same for both verbs, no comma. I let this slide a lot because it often reads better with commas a pauses, but for this case I feel more strongly - with a comma it confuses me a lot when reading since it reads at first like a list of verbs, but the dimension reduction bit is a phrase, not a third verb, so the flow is thrown off.

I'm going to change to this active sentence structure and avoid the whole thing - 
```
The processing pipeline uses `emptyDropsCellRanger()` and `miQC` to filter the raw counts matrix, `scuttle` to log-normalize the counts, and `scater` for dimension reduction.
```

jashapiro commented:
> > and -> Comma
> 
> Only if it's a compound sentence! Since the subject is same for both verbs, no comma. I let this slide a lot because it often reads better with commas a pauses, but for this case I feel more strongly - with a comma it confuses me a lot when reading since it reads at first like a list of verbs, but the dimension reduction bit is a phrase, not a third verb, so the flow is thrown off.

It _should_ be a list of verbs, but it was the two "and"s in close proximity that really bothered me. If only we had a verb for  "dimension reduction". But really, they shouldn't have the same subject, because the normalization isn't done on the raw counts matrix! So a complete rewrite is better all around.

sjspielman commented:
> If only we had a verb for "dimension reduction".

Yes, this is the crux problem! Luckily not a the worst problem in the world :)
